# Profitable App Profiles for the App Store and Google Play Market

For this project we will analize free Android and iOS apps to know what type of them are likely to attract more users.

##  Exploring Data

In this project we use two datasets: 
1. The first one from Google Play about approximately 10,000 Android apps. (2018) [Link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv) 
2. The second one from the App Store about approximately 7,000 iOS apps. (2017) [Link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv)

In [5]:
def explore_data(dataset, start,end,rows_and_columns=False):
    dataset_slice=dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') #Adds a new empty line after each row
        if rows_and_columns:
            print('Number of rows:', len(dataset))
            print('Number of columns',len(dataset[0]))

We created a function to make easier the task to explore the datasets. The function is named `explore_data()`

In [6]:
from csv import reader
#Google Play
opened_file=open('googleplaystore.csv',encoding='utf8') #Open dataset
read_file = reader(opened_file) 
android = list(read_file)
android_header= android[0]
android=android[1:]

#Apple Store
opened_file= open('Applestore.csv',encoding='utf8')
read_file = reader(opened_file)
ios = list(read_file)
ios_header= ios[0]
ios=ios[1:]

In [7]:
print(android_header)
print('\n')
explore_data(android,0,5)
print(len(android[0]))
print(len(android))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

In [8]:
print(ios_header)
print('\n')
explore_data(ios,0,5)
print(len(ios[0]))
print(len(ios))

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


16
7197


 From  this function we can see that the Google Play dataset has 10841 apps and 13 columns, while the AppStore dataset has 7197 and 16 columns.
 The columns that could help us in our analysis are:
* Google Play: 'Category','Reviews','Installs ','Type','Price','Genres'
* AppStore: 'price','user_rating'cont_rating,'prime_genre','rating_count_ver','currency','track_name'

## Data Cleaning

### Deleting wrong data

We will proceed to clean the datasets to make sure that the data is accurate. 

In [9]:
print(android[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [10]:
del android[10472]

In [11]:
for row in ios:
    if len(row)!=len(ios_header):
        print(row) #this will print the row data.
print(ios.index(row)) 

7196


In [12]:
print(ios[7196])

['977965019', 'みんなのお弁当 by クックパッド ~お弁当をレシピ付きで記録・共有~', '51174400', 'USD', '0.0', '0', '0', '0.0', '0.0', '1.4.0', '4+', 'Food & Drink', '37', '0', '1', '1']


In [13]:
del ios[7196]

### Removing Duplicates

There are multiple duplicate rows in both datasets. We will remove them to ensure that there are no unnecesary information.

In [14]:
for app in android:
    name=app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


We will not remove duplicates randomly. Instead, we will only keep the row with the highest number of reviews and remove the other entries for any given app

In [15]:
duplicate_apps=[]
unique_apps=[]
for app in android:
    name=app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print('Nº Duplicates: ',len(duplicate_apps))
print('\n')
print('Example of Duplicate: ',duplicate_apps[:15])

Nº Duplicates:  1181


Example of Duplicate:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We will proceed to create a dictionary, where each dictionary key is a unique app name and the value is the highest number of reviews of that app.

In [16]:
reviews_max={}
for apps in android:
    name = apps[0]
    n_reviews=float(apps[3])
    if name in reviews_max and reviews_max[name]<n_reviews:
        reviews_max[name]=n_reviews
    elif name not in reviews_max:
        reviews_max[name]=n_reviews

In [17]:
print(len(reviews_max))

9659


In [18]:
android_clean=[]
already_added=[]
for apps in android:
    name=apps[0]
    n_reviews=float(apps[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(apps)
        already_added.append(name)

In [19]:
print(len(android_clean))

9659


### Removing Non-English Apps

We will write a function that removes non-english characters because we want to analyze apps towards an English-speaking audicience.
The function will take a string and returns `False` if there's a character in the string that does not belong to commom English characters, otherwise it returns `True`

In [20]:
def english(string):
    non_ascii=0
    for n in string:
        if ord(n) > 127:
            non_ascii+=1
    if non_ascii > 3:
            return False
    else:
        return True

In [21]:
print(english('Instagram'))
print(english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english('Docs To Go™ Free Office Suite'))
print(english('Instachat 😜'))

True
False
True
True


We will use this function to filter out non-English apps from both datasets

In [22]:
android_english=[]
ios_english=[]

for apps in android_clean:
    name=apps[0]
    if english(name):
        android_english.append(apps)
for apps in ios:
    name=apps[1]
    if english(name):
        ios_english.append(apps)


In [23]:
print(len(android_english))
print(len(ios_english))

9614
6183


Removing non-english apps we are left with 9614 apps in the Google Play Store dataset and 6183 apps in the AppStore dataset.

### Removing paid Apps

In [24]:
android_free=[]
ios_free=[]
for apps in android_english:
    free_a=apps[7]
    if free_a == '0':
        android_free.append(apps)
for app in ios_english:
    free_i=app[4]
    if free_i=='0.0':
        ios_free.append(app)


In [25]:
print(len(android_free))
print(len(ios_free))

8864
3222


In the final process of our data cleaning we are left with 8864 Android Apps and 3222 iOS Apps

The main purpose of this project is to determine the kinds of apps that are more likely to attract users

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

* Build a minimal Android version of the app, and add it to Google Play.
* If the app has a good response from users, we develop it further.
* If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

In [26]:
def freq_table(dataset,index):
    genre_dic={}
    total=0
    for row in dataset:
        total +=1
        genre=row[index]
        if genre in genre_dic:
            genre_dic[genre]+=1
        else:
            genre_dic[genre]=1
            
    genre_per={}
    for key in genre_dic:
        percentage =(genre_dic[key]/total)*100
        genre_per[key]=percentage
    return genre_per

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [27]:
display_table(ios_free,11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Analyzing iOS apps we can clearly see that: 
* 'Games' is the most frequent genre in the dataset and 'Entertainment' is second with a difference of past 50%
* The rest of the genres are far less frequent, not even reaching 5%.
* The sum of the rest of the genres does not surpass the first category (41,73%)

From the data above we can see that there are far more apps that are use as a distraction than productivity or educational apps. So, for the Apps Store, games are the apps that could attract more users.

We will proceed to analyze the Google Play Store dataset

In [28]:
display_table(android_free,1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [29]:
display_table(android_free,9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

From this dataset we can analyze that the most common category is 'FAMILY' (18%) and with far less percentage the 'GAME' (9%) category.
The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.
We can see that there are far more categories and are more balanced than the App Store Dataset. To summarize, the difference between the AppStore and the Google Play Store is that market of the former is concentrated on Games while the latter has more distribution of the categories. 

We will now find the genres with most users 

In [30]:
prime_genre=freq_table(ios_free,11)
for genre in prime_genre:
    total=0
    len_genre=0
    for genre_2 in ios_free:
        genre_app=genre_2[11]
        if genre_app == genre:
            n_ratings=float(genre_2[5])
            total +=n_ratings
            len_genre+=1
    avg_n_ratings=total/len_genre
    print(genre,':',avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


We can see that on average, Navigation, Social Networking and Music apps have the highest number of reviews (heavily influenced by Google Maps, that it's an essential app.

In [74]:
category_android=freq_table(android_free,1)
for category in category_android:
    total=0
    len_category=0
    for app in android_free:
        category_app=app[1]
        if category_app==category:
            n_install=app[5]
            n_install=n_install.replace(',','')
            n_install=n_install.replace('+','')
            total+=float(n_install)
            len_category+=1
    avg_n_install=total/len_category
    print(category,':',avg_n_install)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

We can see that Communication apps have the most installs